posts/inside-rust/2024-08-15-this-development-cycle-in-cargo-1.81.md

```diff
@@ -0,0 +1,426 @@
+and a [PR](https://github.com/rust-lang/rust/pull/126930) against the compiler.
+However, to do this with `build.rs` would require `build.rs` files to opt-in and they would need to be coordinate on checksum algorithms.
+We decided to punt on this and leave `build.rs` using mtime for now.
+
+[Xaeroxe](https://github.com/Xaeroxe) posted [#14137](https://github.com/rust-lang/cargo/pull/14137) an initial version of checksum fingerprinting for Cargo.
+
+##### `cargo info`
+
+*Update from [1.79](https://blog.rust-lang.org/inside-rust/2024/05/07/this-development-cycle-in-cargo-1.79.html#cargo-info)*
+
+After waiting some time to collect feedback,
+[Rustin170506](https://github.com/Rustin170506)
+proposed `cargo info` to be merged into cargo ([#14141](https://github.com/rust-lang/cargo/pull/14141)).
+This included adapting the code from being written as a third-party to being written in a way Cargo expects, documentation was added, and completion scripts.
+
+There was some discussion on whether to insta-stabilize.
+As a third-party command, we are able to get the testing benefits of nightly-only features.
+If we don't insta-stabilize it, you could only run it as `cargo +nightly -Zsubcommand info` or run the existing `cargo-info` as `cargo info` but `cargo-info info`; `cargo info` would fail.
+Like with `cargo add` and `cargo remove`, we decided to insta-stabilize.
+
+FCP was started for this command.
+
+## Design discussions
+
+##### `--lockfile-path`
+
+When you (or rust-analyzer) runs `cargo metadata`,
+Cargo will ensure the lockfile is in-sync with the `Cargo.toml` file,
+generating it if needed.
+However, if you do this on a read-only filesystem, then it fails
+([#10096](https://github.com/rust-lang/cargo/issues/10096)).
+
+When discussing that problem, the main perspectives were
+
+- `cargo metadata` should be consistent in its output which can only happen if we can write a `Cargo.lock` to the filesystem
+- `cargo metadata` should always let me query information, even it means the information changes due to outside changes (e.g. a new version of a dependency is published so it gets selected)
+
+The problem I see with the latter is user intention.
+Today, you can run `cargo metadata --no-deps` and no `Cargo.lock` gets generated so there isn't a problem.
+Once you ask for dependencies,
+you are asking for dependencies of that instance of the project and an ephemeral state. 
+
+<!-- 2024-07-16 -->
+One proposed compromise was a
+[`--lockfile-path`](https://github.com/rust-lang/cargo/issues/5707)
+flag that allowed callers to move the lockfile generation to a writeable location.
+[Ifropc](https://github.com/Ifropc) drove the conversation on this and we discussed it further amonmg the Cargo team.
+
+For the design, we decided to be consistent in behavior with `--manifest-path`
+- CLI flag, not in the manifest or in config
+- `lockfile_path.file_name() == "Cargo.lock"` must be true
```

Comment:
One question I had about this is whether `--lockfile-path` is purely write-only, or if it's possible to use `--lockfile-path` to specify a different lockfile for cargo operations that will use it in a read-only fashion.

Perhaps this isn't decided yet and I'm jumping ahead of the curve here. But I figured it could be worth a brief mention one way or the other.

Reply:
It is an alternative path. Could also read from that path.

So in practice you can provide a path to `Cargo.msrv.lock` with it.
(That might not be needed when MSRV-aware resolver becomes default.)

Reply:
Nice! 🎉

Lots of great stuff in this release, and in 1.81 as a whole. Super excited for it! 🤩

Reply:
78b29685 clarified the intention

> So in practice you can provide a path to Cargo.msrv.lock with it.
(That might not be needed when MSRV-aware resolver becomes default.)

That is currently blocked because we require `Cargo.lock` to be the `file_name`.  Someone would have to do `msrv/Cargo.lock`

Reply:
New wording looks good!